New Jobs UIJobInfo Dialog / Alert

thertzelle opened the pull request:
This PR adds a new UIJobInfo class that is able to take in variables from both the UIPersonPage and UILotControl (Job Grade [Level], Job Type, and Total Cumulative Experience Points), and display a nicer looking UI and provide some additional information that is helpful as you progress through your job. The two new bits of information is what you current job level is, and your performance, which is a progress bar, that gives you a % till promotion.

These changes have been extensively tested on all the current job career paths and tested on various roll overs from one grade to another.

This is my first UI element addition to FreeSO, so I did my best to make it look as nice as I could, and match the required look and feel of FreeSO / TSO. Please let me know here if you have any additional feedback or suggestions; or find me on Discord.

Here is how the current dialog looks when you retrieve it by phone call:
<img width="286" alt="Current Dialog" src="https://user-images.githubusercontent.com/6549835/54869912-2e980700-4d6d-11e9-8f0f-d753f89ac235.PNG">

And here is the new version of the dialog (called with a different sim):
<img width="317" alt="New Dialog" src="https://user-images.githubusercontent.com/6549835/54869924-4d969900-4d6d-11e9-985a-49bf17e7789d.PNG">

Currently for this update, if you are unemployed, you will still see the previous style dialog:
<img width="275" alt="Current Unemployed Dialog" src="https://user-images.githubusercontent.com/6549835/54869928-5d15e200-4d6d-11e9-9b87-ddb344c85def.PNG">

Here are a few more additional screenshots with different job grades and job types:
<img width="324" alt="Extra Screenshot 1" src="https://user-images.githubusercontent.com/6549835/54869932-7cad0a80-4d6d-11e9-9f46-3283c21c0c37.PNG">
<img width="315" alt="Extra Screenshot 2" src="https://user-images.githubusercontent.com/6549835/54869935-846caf00-4d6d-11e9-9834-7eb3a0eafed8.PNG">
<img width="314" alt="Extra Screenshot 3" src="https://user-images.githubusercontent.com/6549835/54869937-964e5200-4d6d-11e9-8348-f5ceffaef337.PNG">
<img width="317" alt="Extra Screenshot 4" src="https://user-images.githubusercontent.com/6549835/54869941-afef9980-4d6d-11e9-802d-1c6442e1a732.PNG">

- Jones

riperiperi commented on TSOClient/tso.client/UI/Panels/UILotControl.cs:
```diff
@@ -280,6 +281,11 @@ void vm_OnDialog(FSO.SimAntics.Model.VMDialogInfo info)
 
             VMDialogType type = (info.Operand == null) ? VMDialogType.Message : info.Operand.Type;
 
+            if (info.Operand.MessageStringID == 19)
```
This is a hack. You must edit phoneglobals.iff with volcanic and change the dialog primitive type to this instead. (and include the final result as a piff)

thertzelle replied:
Removed this hack and will include an updated `jobcontroller.piff` which changes the Operand Type for this Dialog to "FSOJob" /  127 per the enum.

riperiperi commented on TSOClient/tso.client/UI/Panels/UILotControl.cs:
```diff
@@ -322,6 +328,19 @@ void vm_OnDialog(FSO.SimAntics.Model.VMDialogInfo info)
+                        JobInformation JobInfo = new JobInformation((int)jobInfo.Value.Level, (int)jobInfo.Key, (int)jobInfo.Value.Experience); //Job Grade, Job Type, Experience
+                        var jobInfoAlert = new UIJobInfo(JobInfo);
+                        jobInfoAlert.Show();
+                    }
```
This is not tied into the rest of the dialog system. It's fine if the dialog does not block (if you haven't encountered any issues, it probably doesn't) but in future please try to make your dialog modifications a GenericAddition so that the blocking dialog stuff still works.

thertzelle replied:
I will keep this in mind for any future changes and might refactor as I become a bit more comfortable in the future, I have no had any blocking issues during testing.

riperiperi commented on TSOClient/tso.client/UI/Profile/UIJobInfo.cs:
```diff
@@ -0,0 +1,377 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace FSO.Client.UI.Profile
```
Did you mean "model"? This should go in FSO.Client.UI.Panels, or some subfolder in there.

thertzelle replied:
I updated the using declarations to a more appropriate; refined path.

riperiperi commented on TSOClient/tso.client/UI/Profile/UIJobInfo.cs:
```diff
@@ -0,0 +1,377 @@
+        public override void Removed()
+        {
+            base.Removed();
+            Icon.Texture?.Dispose();
```
Where is Icon defined? Is it ever set?

thertzelle replied:
Removed Icon Dispose; Replaced with Dispose for bespoke texture for progress bar.

riperiperi commented on TSOClient/tso.client/UI/Profile/UIJobInfo.cs:
```diff
@@ -0,0 +1,377 @@
+                NextPosition.Caption = jobInformation.NextPosition;
+                ProgressBar.Value = jobInformation.PromotionPercentage;
+            }
+            ShowDialog(this, false);
```
UIScreen.GlobalShowDialog(). It is a static function so you can call it without any reference.

thertzelle replied:
Removed my copied function and now calling UIScreen...Dialog()

riperiperi commented on TSOClient/tso.client/UI/Profile/UIJobInfo.cs:
```diff
@@ -0,0 +1,377 @@
+            this.CenterAround(UIScreen.Current, -(int)UIScreen.Current.X * 2, -(int)UIScreen.Current.Y * 2);
+        }
+
+        public void Destory()
```
Destroy

thertzelle replied:
Renamed Spelling Mistake.

riperiperi commented on TSOClient/tso.client/UI/Profile/UIJobInfo.cs:
```diff
@@ -0,0 +1,377 @@
+                ProgressBar.SetSize(progressBarWidth, 27);
+                ProgressBar.Caption = "";
+                
+                ProgressBar.Background = TextureGenerator.GenerateRoundedRectangle(GameFacade.GraphicsDevice, new Color(37, 58, 77), progressBarWidth, 27, 9);
```
This is a bespoke texture, so perhaps you should store a private reference to it that you can `.Dispose()` during the `OnRemoved()` override below.

thertzelle replied:
Calling `.Dispose()` on `ProgressBar.Background.Dispose()` during the Remove event. I hope this accomplishes the same goal.

riperiperi commented on TSOClient/tso.client/UI/Profile/UIJobInfo.cs:
```diff
@@ -0,0 +1,377 @@
+                CaptionStyle = TextStyle.DefaultTitle
+            };
+            HoursTitle.CaptionStyle = HoursTitle.CaptionStyle.Clone();
+            HoursTitle.CaptionStyle.Color = new Color(238, 247, 169);
```
No need setting default values - the yellow colour and shadow false are default, so they do not need to be reset when cloning the default title.

thertzelle replied:
Removed setting unnecessary properties on yellow text non-shadow items.

thertzelle commented:
Updated this branch and PR with a new commit to fix the requested issues: https://github.com/riperiperi/FreeSO/pull/136/commits/6ccbc35371050f31283a3cae867c663846467714

- **FSOFacade**: Updating the location of the profile, allowed us to use the previous 'using FSO.Client.UI.Panels' declaration.
- **UILotControl**: Removed hack for grabbing MessageStringID 19. Updated jobcontroller.piff to set the type from message to FSOJOB.
- **UIPersonPage**: Updated the using directive location to be more appropriate.
- **UIJobInfo**: Updated the namespace of location of this class; Corrected spelling mistake of Destroy. Removed default color and shadow sets on yellow text fields; calling Dispose on bespoke texture for progress bar. Remove redeleration of 'ShowDialog' in favor of using the UIScreen.GloablShowDialog method.
- **VMDialogPrivateStrings**: Set this to 127 to prove that I knew what I was modifying, and just left it here. It IS set to 127 in the included piff change.

Here is the attached jobcontroller.piff file, minor change was to the 4096 - Get Current Job Info Tree, Used by all Phones for "Call Work". Dialog Type was changed to FSOJob / 127 from the `VMDialogType` enum.

[jobcontroller.zip](https://github.com/riperiperi/FreeSO/files/3006327/jobcontroller.zip) - File had to be zipped as github doesn't accept .piff files; If you would be more comfortable I can sent you the file in discord.
